exercises/Robert_Applin/ex01_basics/src/main.cpp

```diff
@@ -0,0 +1,203 @@
+/**
+ * Skeleton main routine
+ */
+
+#include <algorithm>
+#include <cctype>
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <iterator>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+using WordCount = std::vector<std::pair<std::string, std::size_t>>;
+
+/**
+ * @return An iterator pointing to the next letter contained
+ *         in the supplied set.
+ */
+std::string::const_iterator
+findNextCharacter(std::string::const_iterator fromIterator,
+                  std::string::const_iterator toIterator,
+                  const std::string &characters) {
+  for (auto it = fromIterator; it < toIterator; ++it) {
+    if (characters.find(*it) != std::string::npos)
+      return it;
+  }
+  return toIterator;
+}
+
+/**
+ * @return An iterator pointing to the next letter not
+ *         contained in the supplied set.
+ */
+std::string::const_iterator
+findNextNotCharacter(std::string::const_iterator fromIterator,
+                     std::string::const_iterator toIterator,
+                     const std::string &characters) {
+  for (auto it = fromIterator; it < toIterator; ++it) {
+    if (characters.find(*it) == std::string::npos)
+      return it;
+  }
+  return toIterator;
+}
+
+/**
+ * @return A vector of sub strings split around delimiters
+ */
+std::vector<std::string> splitBy(std::string::const_iterator fromIterator,
+                                 std::string::const_iterator toIterator,
+                                 const std::string &delimiters) {
+  std::vector<std::string> substrings;
+  auto iter = findNextNotCharacter(fromIterator, toIterator, delimiters);
+
+  while (iter != toIterator) {
+    const auto nextIter = findNextCharacter(iter, toIterator, delimiters);
+    substrings.emplace_back(iter, nextIter);
+    iter = findNextNotCharacter(nextIter, toIterator, delimiters);
+  }
+  return substrings;
+}
+
+/**
+ * @return A true if count1 is larger than count2
+ */
+bool compareCounts(const std::pair<std::string, std::size_t> &count1,
+                   const std::pair<std::string, std::size_t> &count2) {
+  return count1.second > count2.second;
+}
+
+/**
+ * @return True if the string is shorter than 5 characters
+ */
+bool isShorterThanFive(const std::string &str) { return str.length() < 5; }
+
+/**
+ * Removes any words shorter than five from the vector
+ */
+void removeShorterThanFive(std::vector<std::string> &strings) {
+  auto it = std::remove_if(strings.begin(), strings.end(), isShorterThanFive);
+  // Remove invalid indices
+  strings.erase(it, strings.end());
+}
+
+/**
+ * @return A string transformed to lower case
+ */
+std::string toLower(std::string str) {
+  std::transform(str.begin(), str.end(), str.begin(),
+                 static_cast<int (*)(int)>(std::tolower));
```

Comment:
Prefer a lambda over this weird cast to a function pointer:
```cpp
 std::transform(str.begin(), str.end(), str.begin(), [](const char c){ std::tolower(c); });
```

It should equate to the same thing in any decent compiler.

Reply:
That looks much better!